lib/util/result.ts

```diff
@@ -1,73 +1,151 @@
-      (error) => Result.err(error)
-    );
-  }
+  private constructor(private readonly res: Res<T, E>) {}
 
-  static wrap<T>(callback: () => T): Result<T>;
-  static wrap<T>(promise: Promise<T>): Promise<Result<T>>;
-  static wrap<T>(
-    input: (() => T) | Promise<T>
-  ): Result<T> | Promise<Result<T>> {
-    return input instanceof Promise
-      ? Result.wrapPromise(input)
-      : Result.wrapCallback(input);
+  unwrap(): Res<T, E>;
+  unwrap<U>(fallback: U): T | U;
+  unwrap<U>(fallback?: U): Res<T, E> | T | U {
+    if (arguments.length === 0) {
+      return this.res;
+    }
+    return this.res.ok ? this.res.value : (fallback as U);
   }
 
-  private constructor(public readonly res: Res<T, E>) {}
+  transform<U, EE>(fn: (value: T) => Result<U, EE>): Result<U, E | EE>;
+  transform<U, EE>(
+    fn: (value: T) => Promise<Result<U, EE>>
+  ): AsyncResult<U, E | EE>;
+  transform<U, EE>(
+    fn: (value: T) => Result<U, EE> | Promise<Result<U, EE>>
+  ): Result<U, E | EE> | AsyncResult<U, E | EE> {
+    if (!this.res.ok) {
+      return Result.err(this.res.error);
+    }
+
+    try {
+      const res = fn(this.res.value);
+
+      if (res instanceof Promise) {
+        return new AsyncResult((resolve) => {
+          res.then(resolve).catch((error) => {
+            logger.warn(
+              { err: error },
+              'Result: unhandled async transform error'
+            );
+            resolve(Result.err(error) as never);
+          });
+        });
+      }
 
-  transform<U>(fn: (value: T) => U): Result<U, E> {
-    return this.res.success
-      ? Result.ok(fn(this.res.value))
-      : Result.err(this.res.error);
+      return res;
+    } catch (error) {
+      logger.warn({ err: error }, 'Result: unhandled transform error');
+      return Result.err(error);
+    }
   }
+}
 
-  catch<U>(fallback: U): T | U {
-    return this.res.success ? this.res.value : fallback;
+export class AsyncResult<T, E> extends Promise<Result<T, E>> {
+  constructor(
+    executor: (
+      resolve: (value: Result<T, E> | PromiseLike<Result<T, E>>) => void,
+      reject: (reason?: unknown) => void
+    ) => void
+  ) {
+    super(executor);
   }
 
-  get value(): T | undefined {
-    return this.res.success ? this.res.value : undefined;
+  static wrap<T, E = Error>(promise: Promise<T>): AsyncResult<T, E> {
+    return new AsyncResult((resolve) => {
+      promise
+        .then((value) => resolve(Result.ok(value)))
+        .catch((error) => resolve(Result.err(error)));
+    });
   }
 
-  get error(): E | undefined {
-    return this.res.success ? undefined : this.res.error;
+  unwrap(): Promise<Res<T, E>>;
+  unwrap<U>(fallback: U): Promise<T | U>;
+  unwrap<U>(fallback?: U): Promise<Res<T, E> | (T | U)> {
+    return arguments.length === 0
+      ? this.then((res) => res.unwrap())
+      : this.then((res) => res.unwrap(fallback as U));
+  }
+
+  transform<U, EE>(fn: (value: T) => Result<U, EE>): AsyncResult<U, E | EE>;
```

Comment:
why it can still be `E` instead of only `EE`? 🤔

Reply:
Because `E` is the class-level variable that is already being used. I wasn't able to come up with good name for this one 🙂

`EE` meant to signify errors possible during this particular transform, so they're form union type together with `E` which is from `AsyncResult<T, E>`.

Reply:
ok